### DISCUSSION TITLE
Stubgen does not generate stubs for implicit constructions

### DISCUSSION THREAD

**Discussion by efifogel** [wjakob/nanobind#684] [General]
Assume you have a class A with an implicit constructor from double and a class B with a constructor from A.
The binding code looks like this:
```
nb::class_<A>(m, "A").def(nb::init_implicit<double>())...
nb::class_<B>(m, "B").def(py::init<const A&>())...
```
In python we can write

b = B(1.234)

However, the generated pyi file has only a stub for the constructor of B from A. 
As a consequence, type checking fails.
I know I can explicitly add a signature:
```
nb::class_<B>(m, "B").def(py::init<const A&>(), nb::sig("def __init__(self, x: A | float) -> None")))...
```
Is there a better way to handle this?

**Comment by wjakob** [comment]
Implicit conversions are a somewhat advanced feature. The Python typing infrastructure AFAIK doesn't support a way of specifying them except for copy-pasting unions all possible conversions all over the place (which is unfortunate).

nanobind's stubgen currently doesn't provide a way of performing this transformation automatically in generated stubs, and it would be complicated to realize such a feature. My suggestion would be that you either override signatures via `nb::sig` or more systematically using a pattern file (or perhaps even a custom regular expression or sed script that runs over the generated signature file).
